core-client/src/transports/duplex.rs

```diff
@@ -12,27 +13,61 @@ use std::collections::VecDeque;
 use super::RequestBuilder;
 use crate::{RpcChannel, RpcError, RpcMessage};
 
+struct Subscription {
+	/// Subscription id received when subscribing.
+	id: Option<SubscriptionId>,
+	/// Rpc method to unsubscribe.
+	unsubscribe: String,
+	/// Where to send messages to.
+	channel: mpsc::Sender<Result<Value, RpcError>>,
+}
+
+impl Subscription {
+	fn new(channel: mpsc::Sender<Result<Value, RpcError>>, unsubscribe: String) -> Self {
+		Subscription {
+			id: None,
+			unsubscribe,
+			channel,
+		}
+	}
+}
+
 /// The Duplex handles sending and receiving asynchronous
 /// messages through an underlying transport.
 pub struct Duplex<TSink, TStream> {
 	request_builder: RequestBuilder,
-	queue: HashMap<Id, oneshot::Sender<Result<Value, RpcError>>>,
-	sink: TSink,
-	stream: TStream,
+	/// Channel from the client.
 	channel: Option<mpsc::Receiver<RpcMessage>>,
+	/// Requests that haven't received a response yet.
+	pending_requests: HashMap<Id, oneshot::Sender<Result<Value, RpcError>>>,
+	/// Subscription requests that haven't received a subscription id yet.
+	pending_subscriptions: HashMap<Id, String>,
+	/// A map from the subscription name to the subscription.
+	subscriptions: HashMap<String, Subscription>,
```

Comment:
I don't follow, why would a client subscribe multiple times to the same notification?

The reason it is done like this is because we don't know the Id yet when creating a subscription, until the server assigns an id. The only thing we do know is the name of the subscription, so where do we put the mpsc channel in the meantime?

Reply:
Opened an issue for this

Reply:
Linking the issue for posterity: https://github.com/paritytech/jsonrpc/issues/444

We don't know the subscription Id, but we know the request id that user set (I mean the `{jsonrpc:2.0, id:XXX, method:...}` field of the jsonrpc request.) - the subscription response will contain that as well and we should use it to match one and the other.